cpg_utils/flows/metamist.py

```diff
@@ -0,0 +1,537 @@
+                )
+                found_output_fpath = None
+            elif not exists(found_output_fpath):
+                logging.error(
+                    f'Found a completed analysis {label}, '
+                    f'but the "output" file {found_output_fpath} does not exist'
+                )
+                found_output_fpath = None
+
+        # completed and good exists, can reuse
+        if found_output_fpath:
+            logging.info(
+                f'Completed analysis {label} exists, '
+                f'reusing the result {found_output_fpath}'
+            )
+            return found_output_fpath
+
+        # can't reuse, need to invalidate
+        if completed_analysis:
+            logging.warning(
+                f'Invalidating the analysis {label} by setting the status to "failure", '
+                f'and resubmitting the analysis.'
+            )
+            self.update_analysis(completed_analysis, status=AnalysisStatus.FAILED)
+
+        # can reuse, need to create a completed one?
+        if exists(expected_output_fpath):
+            logging.info(
+                f'Output file {expected_output_fpath} already exists, so creating '
+                f'an analysis {label} with status=completed'
+            )
+            self.create_analysis(
+                type_=analysis_type,
+                output=expected_output_fpath,
+                status='completed',
+                sample_ids=sample_ids,
+                dataset=dataset or self.default_dataset,
+            )
+            return expected_output_fpath
+
+        # proceeding with the standard workflow (creating status=queued, submitting jobs)
+        else:
+            logging.info(
+                f'Expected output file {expected_output_fpath} does not exist, '
+                f'so queueing analysis {label}'
+            )
+            return None
+
+    def get_ped_entries(self, dataset: str | None = None) -> list[dict[str, str]]:
+        """
+        Retrieve PED lines for a specified SM project, with external participant IDs.
+        """
+        dataset = dataset or self.default_dataset
+
+        families = self.fapi.get_families(dataset)
+        family_ids = [family['id'] for family in families]
+        ped_entries = self.fapi.get_pedigree(
+            internal_family_ids=family_ids,
+            response_type='json',
+            project=dataset,
+            replace_with_participant_external_ids=True,
+        )
+
+        return ped_entries
+
+
+@dataclass
+class MmSequence:
+    """
+    Metamist "Sequence" entry.
+
+    See metamist for more details:
+    https://github.com/populationgenomics/sample-metadata
+    """
+
+    id: str
+    sample_id: str
+    meta: dict
+    sequencing_type: str
+    alignment_input: AlignmentInput | None = None
+
+    @staticmethod
+    def parse(data: dict, check_existence: bool = False) -> 'MmSequence':
+        """
+        Parse dictionary to create a SmSequence object.
```

Comment:
The mix between `Mm` and `Sm` is a little confusing, but there's probably no great way around this until everything has been renamed on the Metamist side.

Reply:
Renamed to simply `Sequence`